### PR TITLE
Support nesting block comments in CI grammar

### PR DESCRIPTION
**Pull request by mehmetoguzderin**
For a resolved discussion. See: https://github.com/gpuweb/gpuweb/pull/2288

### CODE REVIEW

File: wgsl/extract-grammar.py [hunk 1]
```diff
@@ -419,6 +425,78 @@ def not_token_only(value):
     grammar_package.write('    "main": "bindings/node"\n')
     grammar_package.write('}\n')
 
+# External scanner for nested block comments
+# See: https://github.com/tree-sitter/tree-sitter-rust/blob/master/src/scanner.c
+
```
Comment (dneto0) [inline wgsl/extract-grammar.py:430]: Nitpick:  I suggest also referencing https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners
That explains the API in detail.
(Except it doesn't explain the 'reset' method.  oops. :-)  )

File: wgsl/extract-grammar.py [hunk 2]
```diff
@@ -419,6 +425,78 @@ def not_token_only(value):
...
+  if (lexer->lookahead == '/') {
+    advance(lexer);
+    if (lexer->lookahead != '*') return false;
+    advance(lexer);
+
+    bool after_star = false;
+    unsigned nesting_depth = 1;
+    for (;;) {
+      switch (lexer->lookahead) {
+        case '\0':
```
Comment (dneto0) [inline wgsl/extract-grammar.py:464]: So this is interesting.  I had not thought about the NUL character in source code.
UTF-8 allows it.  We would need a decision in the group about whether WGSL source can contain it.

I'll file an issue to track that.
I'll approve this and we can revisit if the group decides to allow NUL characters in comments.

**Review by dneto0 — APPROVED** [review]
Thanks!
This will allow us to add good block comment examples to the spec.

File: wgsl/extract-grammar.py (hunk 1 above)
Comment (mehmetoguzderin) [reply wgsl/extract-grammar.py:430]: @dneto0 definitely makes sense, tried here: https://github.com/gpuweb/gpuweb/pull/2292

File: wgsl/extract-grammar.py (hunk 2 above)
Comment (mehmetoguzderin) [reply wgsl/extract-grammar.py:464]: @dneto0 thanks for pointing it out, a spec discussion will definitely benefit.

File: wgsl/extract-grammar.py (hunk 1 above)
Comment (dneto0) [reply wgsl/extract-grammar.py:430]: Interesting.  The docs are more up to date than the Rust scanner.
The reset method can be removed.

Comment (mehmetoguzderin) [reply wgsl/extract-grammar.py:430]: @dneto0 indeed that builds fine, should have checked in the first place...: https://github.com/gpuweb/gpuweb/pull/2293

File: wgsl/extract-grammar.py (hunk 2 above)
Comment (dneto0) [reply wgsl/extract-grammar.py:464]: I just read tree-sitter's lexer.   I  think this actually signals the end of input.

Internally, when the lexer hits end of input, it sets TSLexer.lookahead to '\0' and the internal "lookahead_size" to 1.
See https://github.com/tree-sitter/tree-sitter/blob/67de9435b109f9bd8bf5957d11ff13161966d262/lib/src/lexer.c#L150  

This doesn't seem to be documented.  :-(

**Pull request merged**